### PR TITLE
Add a module to handle getups and falling using the Director

### PR DESCRIPTION
**Pull request by TrentHouliston**
This PR adds a module to handle getups

Since Odometry has not always proven the most reliable this is instead using raw sensors with a simple exponential filter and a slightly different method for working out falling and when to get up

Since I can't test this I have no idea if it works!

### CODE REVIEW

File: module/planning/FailureManagement/src/FailureManagement.hpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace module::planning {
+
+    class FailureManagement : public ::extension::behaviour::BehaviourReactor {
+    private:
+        /// @brief Stores configuration values
+        struct Config {
+            struct {
+                struct {
+                    struct {
```
Comment (ysims) [inline module/planning/FailureManagement/src/FailureManagement.hpp:16]: this looks a bit yikes. also initialise variables.

File: module/planning/FailureManagement/src/FailureManagement.hpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                double acc_recovery;
+            } getup;
+        } cfg;
+
+    public:
+        /// @brief Called by the powerplant to build and setup the FailureManagement reactor.
+        explicit FailureManagement(std::unique_ptr<NUClear::Environment> environment);
+
+    private:
+        bool getting_up = false;
```
Comment (ysims) [inline module/planning/FailureManagement/src/FailureManagement.hpp:50]: documentation directives?

File: module/planning/FailureManagement/src/FailureManagement.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                        bool falling = (gyro_mag_state == State::FALLING || acc_mag_state == State::FALLING
+                                        || acc_angle_state == State::FALLING)
+                                       || (gyro_mag_state == State::UNSTABLE && acc_mag_state == State::UNSTABLE)
+                                       || (gyro_mag_state == State::UNSTABLE && acc_angle_state == State::UNSTABLE)
+                                       || (acc_mag_state == State::UNSTABLE && acc_angle_state == State::UNSTABLE);
+
+                        // We are falling
+                        if (falling) {
+                            // TODO emit stability falling
+                            emit(std::make_unique<Stability>(Stability::FALLING));
```
Comment (ysims) [inline module/planning/FailureManagement/src/FailureManagement.cpp:101]: I tthink you've done this todo

File: module/planning/FailureManagement/src/FailureManagement.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            cfg.falling.acc.angle.unstable  = config["falling"]["acc"]["angle"]["unstable"].as<Expression>();
+            cfg.falling.acc.angle.falling   = config["falling"]["acc"]["angle"]["falling"].as<Expression>();
+            cfg.falling.acc.angle.smoothing = config["falling"]["acc"]["angle"]["smoothing"].as<Expression>();
+
+            cfg.getup.angle         = config["getup"]["angle"].as<Expression>();
+            cfg.getup.gyro_recovery = config["getup"]["gyro_recovery"].as<Expression>();
+            cfg.getup.acc_recovery  = config["getup"]["acc_recovery"].as<Expression>();
+        });
+
+        on<Provide<Recover>, Trigger<Sensors>>().then([this](const RunInfo& info, const Sensors& sensors) {
```
Comment (ysims) [inline module/planning/FailureManagement/src/FailureManagement.cpp:56]: should this not be root level?

Comment (ysims) [reply module/planning/FailureManagement/src/FailureManagement.cpp:56]: `Recover` indicates something has gone wrong, but this is more just watching for something to go wrong.

**Review by ysims — COMMENTED** [review]
I don't like the name "FailureManagement". Lots of things can be 'failures'. Maybe "FallManagement" or "FallPlanner".
Odometry currently works for get up detection, we shouldn't regress to using raw sensors. It just wasn't fast enough for falling relax.

**Comment by TrentHouliston** [comment]
This PR has been superseded by a few others that break up this module into 4 others (GetUp skill, GetUpPlanner FallingRelaxPlanner and RecoveryStrategy)

**Pull request closed without merging**